### DISCUSSION TITLE
Swap Tokens with GoSwap with web3

### DISCUSSION THREAD

**Discussion by er1202** [gochain-io/docs#106] [Q&A] [answered]
Hello,

i'm trying to swap some tokens over GoSwap with web3.js or web3.py. As i have seen does gochain web3 not have a function like this.
Is there a "routing contract" involved like on uniswap or how could i do this. I think i have to swap GO's to WGO's first but i can't find a pair for that.

I'm sure the solution is somewhere [here](https://github.com/goswap) but i need a few example code lines with which i can start

Maybe anyone can help me - thank you...

**Comment by er1202** [comment]
I think i found a few lines of sample code but i don't know the **address of the router contract**. Where can i find this.

**Reply by treeder** [reply]
Find it yet?

**Reply by er1202** [reply]
Thank you with your help i got the correct address for the router contract and it works perfect for me.

I also tried to make a few simple swaps with web3 what worked without problems.
The only problem is that i'm only able to swap to WGO (for instance OMI->WGO) and i'm not able to convert WGO->GO.
I am not sure if i have the correct GO address : 0xe8D71132Cd78146fbeD8c085f6c06C - i can't confirm this address on GoChainExplorer so i have 2 addition questions (sorry for that):
1. Is this address the correct GO address for Go's
2. How could i handle it to convert WGO to GO (or GO to WGO) over a CLI or WEB3.js

**Reply by treeder** [reply]
I think this is the WGO contract: https://explorer.gochain.io/addr/0xcC237fa0A4B80bA47992d102352572Db7b96A6B5

And it's standard one that is used for WETH https://github.com/gnosis/canonical-weth/blob/master/contracts/WETH9.sol

**Reply by er1202** [reply]
Thanks for your answer,

i'm able to unwrap WGO to GO with the withdraw function - so far so good but could i wrap GO to WGO, is this one of the functions like transfer, the deposit function seems to be for something else...

**Reply by treeder** [reply]
`deposit` function which is also the default function, so if you simply send GO to the contract address, it will issue you WGO.

**Reply by treeder** [reply]
That work?

**Reply by er1202** [reply]
Yes it does... So wrapping, unwrapping, and shaping from OMI to WGO works perfekt...
Hopefully the last question in this matter is to SWAP from OMI to WGO...
I use: swapExactTokensForTokens(amountIn, amountOutMin, [OmiTokenAddress, WGoTokenAddress]),accountaddress, timeValid) for swapping OMI to WGO but it doesn't work the other way: 
WGO to OMI
swapExactTokensForTokens(amountIn, amountOutMin, [WGoTokenAddress, OmiTokenAddress]),accountaddress, timeValid) 
Any idea what's wrong here...

**Reply by er1202** [reply]
I'm still working (or freaking out) on the problem that i'm able to swap from OMI to WGO but not the other direction...
Could this be a problem with fee's or to i have to send some WGO (GO's whatever) to the contract before the swap???
I see only that the transaction is failed on the GoChainExplorer, without any details [0x001d80f8662bb6429a9f1d0f57243c81505f4500677e2001a941c10790cac827](https://explorer.gochain.io/tx/0x001d80f8662bb6429a9f1d0f57243c81505f4500677e2001a941c10790cac827)

This is my python.code:
`swap_txn = router.functions.swapExactTokensForTokens(web3.toWei(int(amount), 'ether'),
                                                             web3.toWei(int(AmountOutMin), 'ether'),
                                                             [WGoTokenAddress,OmiTokenAddress],
                                                             MetaMaskAddress,
                                                             ValidTime).buildTransaction(
            {'nonce': nonce, 'gas': 139071,  'gasPrice': web3.eth.gas_price})

**Reply by treeder** [reply]
Some googling: 

* https://ethereum.stackexchange.com/questions/96847/uniswap-pancakeswap-swap-swapexacttokensfortokens-swapexactethfortokens-failed 
* https://stackoverflow.com/questions/66404029/web3-py-swapexacttokensfortokens-failing-transaction

**Reply by er1202** [reply]
Thanks for your answer, unfortunately i have seen these pages before...
1. I don't think that i have to less gas. As you see in my transaction gas limit is 139,071 gas used is 37,198 OK?
2. I also tried the swapExactETHForTokens functions but it ditn't work (i have OMI, GO and WGO on my account)
3. I'm not sure how i should work with the aprove() function ??????
4. I tried to put the 'from'  field into the buildTransaction with my MetaMask address (correct?) but still don't work

I don't understand that the other direction works without any problems, do you think i need the aprove function (if so, how to use?)

**Reply by er1202** [reply]
approved it in MetaMask -> now it works... (costs me only 10hours ;-)

**Comment by treeder** [comment, marked as answer]
Answers from discussion:

Router: 0x3881e447F439891dC106Da7bca0007B319eeB74D

Wrapped GO (wGO) contract: 0xcC237fa0A4B80bA47992d102352572Db7b96A6B5 (which is the standard [gnosis contract](https://github.com/gnosis/canonical-weth/blob/master/contracts/WETH9.sol))

Be sure to call `approve()` function before attempting swaps.
